### ISSUE TITLE
[RS0041] Odd oblivious reference types diagnostics being generated

### ISSUE THREAD

**Issue by CodeBlanch** [dotnet/roslyn-analyzers#6921]
### Analyzer

**Diagnostic ID**: `RS0041`: Symbol 'GetEnumerator' uses some oblivious reference types

### Analyzer source

`<PackageVersion Include="Microsoft.CodeAnalysis.PublicApiAnalyzers" Version="[3.11.0-beta1.23402.2]" />`

### Describe the bug

I'm getting a `RS0041: Symbol 'GetEnumerator' uses some oblivious reference types` generated which seems strange. May be related to #4040

### Steps To Reproduce

Take these two classes:

```csharp
public class Happy<T>
{
    public Enumerator GetEnumerator()
    {
        return default;
    }

    public struct Enumerator
    {
    }
}

public class Sad<T>
    where T : class
{
    public Enumerator GetEnumerator()
    {
        return default;
    }

    public struct Enumerator
    {
    }
}
```

The only difference is `Sad<T>` has the `where T : class` constraint.

The quick fix tool thing is generating these two entries in my public api files:

```
Demo.Happy<T>.GetEnumerator() -> Demo.Happy<T>.Enumerator
~Demo.Sad<T>.GetEnumerator() -> Demo.Sad<T>.Enumerator
```

The entry for `Sad<T>` causes the `RS0041`.

### Expected behavior

No diagnostics generated.

### Actual behavior

Diagnostics generated.

**Comment by mavasani** [comment]
@jcouv
